### PR TITLE
feat: add checkpoint save/resume for interrupted training

### PR DESCRIPTION
**Pull request by XiaoBinGan**
## What

Adds automatic checkpoint saving and resume support to the pretraining script (`scripts/train_transformer.py`), addressing **#16**.

### `config/config.py`
- `save_every` ??save a checkpoint every N steps (default 1000, 0 = only at end)
- `checkpoint_dir` ??directory for periodic checkpoints (default `checkpoints/`)
- `keep_last_n` ??keep only the N most recent checkpoints to save disk (default 3)

### `scripts/train_transformer.py`
- **Periodic checkpoint saving** ??model weights, optimizer state, losses, RNG state (for reproducible resume)
- **`--resume` flag** ??auto-resume from the latest checkpoint in `checkpoint_dir`
- **`--resume-from PATH` flag** ??resume from a specific checkpoint file
- **Auto-rotation** ??old checkpoints pruned when `keep_last_n > 0`
- **Final checkpoint** saved after training completes, so resume works for post-hoc fine-tuning
- **Progress bar** correctly offset when resuming mid-training
- `weights_only=False` used for `torch.load` since RNG/numpy state requires pickle

## How to test

```bash
# Train with checkpoints every 1000 steps, keep last 3
python scripts/train_transformer.py
# (configurable via config.py: save_every=1000, keep_last_n=3)

# Simulate interruption (Ctrl+C), then resume:
python scripts/train_transformer.py --resume

# Or resume from a specific checkpoint:
python scripts/train_transformer.py --resume-from checkpoints/checkpoint_step_00500.pt
```

## Compatibility

- Backward compatible ??existing configs work without changes (old defaults)
- Works on CPU and GPU
- RNG state restored for deterministic resume

### CODE REVIEW

**Comment by FareedKhan-dev** [comment]
Hi @XiaoBinGan — thank you for this checkpoint PR! It works (I verified train → checkpoint → `--resume` on a 2×L40 GPU).

We received three implementations of the same checkpoint-resume feature (#19, #20, #28) that conflict with one another, so we've consolidated the feature on **#28**. Two reasons it was chosen as the base: it's Python-3.9-safe (this PR's `find_latest_checkpoint(...) -> str | None` annotation raises `TypeError` at import on 3.9, which the repo still targets), and it uses atomic temp-file writes plus a resume round-trip test.

Closing this one in favor of #28 (now merged).

Good news though: your **other** PR #21 (AMP + gradient checkpointing + gradient accumulation) targets a genuinely important problem — the default-config OOM in #5 — and I'm landing that feature. It needs a couple of fixes first (it currently crashes on CPU because the memory-budget/autocast path calls `torch.cuda` unconditionally, and all four memory flags default ON which silently changes behavior/numerics), so I'm fixing those and rebasing onto the refactored trainer, then merging — and I'll credit you on the merged change. Thanks for both contributions! 🙏
